### ISSUE TITLE
Missing "Match arm is unreachable" warning

### ISSUE THREAD

**Issue by anton-trunov** [FuelLabs/sway#5097]
Here is an example that demonstrates the issue. 

```sway
script;

struct S {
    x: bool
}

fn main() {
    let s = S { x: false };
    let _x = match s {
        S { x } | S { x } => { x }
    };
}
```

**Comment by ironcev** [comment]
Hmm, but that catch-all arm is reachable 🤔 It will match any `s`. Or I am missing something?

**Comment by anton-trunov** [comment]
Yes, it's the second one. Rust reports the following:

```
warning: unreachable pattern
 --> src/main.rs:8:15
  |
8 |     S { x } | S { x } => x,
  |               ^^^^^^^
  |
```

**Comment by ironcev** [comment]
Good point! The current analysis checks only the reachability of the whole arms. That's what the exhausting algorithm gives as a result. So strictly speaking, this is not a bug but a feature 🙂

Seriously, if we want to have this check as well, I would rename the title of this issue to get the new feature and a different message e.g. "Unreachable match arm alternative" or something. But I wouldn't mix it with the overall marcg arm reachability. Because in this case the arm is still reachable and the code in the arm can still be executed.

**Comment by anton-trunov** [comment]
Fair enough, renamed it to "Missing warning about an unreachable pattern in `match` statement"
